File: backend/src/infra/EventBus.ts
import {
  Event,
  EventCallback,
  EventData,
  EventPublisher,
  EventSubscriber,
} from "../types/Events";

export class EventBus implements EventPublisher, EventSubscriber {
  private subscribers: Array<{ event: Event; callback: EventCallback }> = [];

  constructor() {}

  publish<T extends Event>(event: T, data: EventData<T>) {
    const fire = this.subscribers.filter((sub) => sub.event === event);

    fire.forEach((sub) => sub.callback(data));
  }

  subscribe<T extends Event>(event: T, callback: EventCallback) {
    this.subscribers.push({ event, callback });
  }
}

File: backend/src/application/types/event_bus.py
from abc import ABC, abstractmethod
from typing import Callable
from .events import Event, T


class EventBus(ABC):
    """Event bus interface for publishing and subscribing to events."""

    @abstractmethod
    def subscribe(
        self,
        event: Event[T],
        callback: Callable[[T], None],
    ) -> None: ...

    @abstractmethod
    def publish(
        self,
        event: Event[T],
        data: T,
    ) -> None: ...
